Issue 797: generating service permissions

marcdejonge opened the pull request:
So, I've started on an implementation that only parses the DS service definitions for now. Can you review if this is what you were expecting?

See discussion here: https://groups.google.com/forum/#!topic/bndtools-dev/Ra7GDT7x_4U

bjhargrave commented:
Have you seen <https://github.com/bndtools/bnd/blob/master/biz.aQute.bndlib/src/aQute/bnd/indexer/analyzers/SCRAnalyzer.java>?

Also, you should probably develop bnd with the latest bndtools builds. 2.4.0.REL is a bit old now :-)

marcdejonge commented:
Hmm, and I thought I had looked for something like that. But of course I didn't look for SCR...

And for bndtools build you need a snapshot of the latest version?

bjhargrave commented:
You can install the latest bndtools builds from this update site: https://bndtools.ci.cloudbees.com/job/bndtools.master/lastSuccessfulBuild/artifact/build/generated/p2/

djencks commented:
That is totally not the approach I thought we were talking about.  I'm going to have the DS stuff write provides and requires service capabilities into the manifest.  Your permission code should read these (not yet generated) service capabilities and write permissions based on them.  Until they are generated by e.g. ds you can test by just writing a manifest with service provides/requires capabilities by hand.

pkriens commented:
@djencks: I think that is the future. Until that time that we have those, reading the XML makes sense. However, if you think you can quickly provide this 'capability' then it might be worth waiting?

marcdejonge commented:
It seems that this is working well. Can you check and merge it?

bjhargrave commented:
I think per 

> Your permission code should read these (not yet generated) service capabilities and write permissions based on them.

you were to redo your changes to use the generated `osgi.service` capabilities and requirements as the source material to generate the permissions info rather then the parsing the DS xml and processing the imports and exports.

https://github.com/bndtools/bnd/pull/822 is now in bnd to generate the `osgi.service` capabilities and requirements for the DS annotations.

marcdejonge commented:
What exactly is the advantage of parsing the capabilities in stead of the
source XML? I'd rather use the source and this is already working.

There could be an advantage to have an internal list of services that both
these outputs would be based on, but I don't see that in the implementation.
On May 1, 2015 10:41 PM, "BJ Hargrave" <notifications@github.com> wrote:

> I think per
>
> Your permission code should read these (not yet generated) service
> capabilities and write permissions based on them.
>
> you were to redo your changes to use the generated osgi.service
> capabilities and requirements as the source material to generate the
> permissions info rather then the parsing the DS xml and processing the
> imports and exports.
>
> #822 <https://github.com/bndtools/bnd/pull/822> is now in bnd to generate
> the osgi.service capabilities and requirements for the DS annotations.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bndtools/bnd/pull/809#issuecomment-98231004>.
>

bjhargrave commented:
> What exactly is the advantage of parsing the capabilities in stead of the source XML? I'd rather use the source and this is already working. There could be an advantage to have an internal list of services that both these outputs would be based on, but I don't see that in the implementation.

The advantage is that if new features are added to the DS XML, or if the bundle programmer doesn't use DS but writes (or has written) her own osgi.service capabilities and requirements, then by processing the osgi.service capabilities and requirements, this features would work for them. It is future proofing.

djencks commented:
After thinking about it a bit, we decided that the service capabilities/requirements _were_ an appropriate universal store for this information, and any other data structure would be redundant and more likely to cause problems than provide a simplification.

thanks
david jencks

On May 2, 2015, at 5:46 AM, BJ Hargrave <notifications@github.com> wrote:

> What exactly is the advantage of parsing the capabilities in stead of the source XML? I'd rather use the source and this is already working. There could be an advantage to have an internal list of services that both these outputs would be based on, but I don't see that in the implementation.
> 
> The advantage is that if new features are added to the DS XML, or if the bundle programmer doesn't use DS but writes (or has written) her own osgi.service capabilities and requirements, then by processing the osgi.service capabilities and requirements, this features would work for them. It is future proofing.
> 
> —
> Reply to this email directly or view it on GitHub.
>

marcdejonge commented:
You have a good point and it does seem reasonable to use the capabilities as a generic storage.

Is there then also some intermediate representation in the builder for the generated capabilities? Or should I just parse the properties?

bjhargrave commented:
You need to remove the merge commits. Your PR must not include any merges, just plain commits.

marcdejonge commented:
Ok, just rebased and added another feature. Hopefully this will built correctly.

marcdejonge commented:
And should I also update the documentation about this? Can I just edit this file:

https://github.com/bndtools/bnd.manual/blob/gh-pages/_macros/permissions.md

???

bjhargrave commented:
> And should I also update the documentation about this? Can I just edit this file:
> 
> https://github.com/bndtools/bnd.manual/blob/gh-pages/_macros/permissions.md
> 
> ???

Once the PR is merged, then the page on the gh-pages branch of the bnd.manual repo is the proper place to update the documentation.

marcdejonge commented:
I've tested it and it seems to work perfectly. Can you look at it and merge it?

bjhargrave commented on biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java:
```diff
@@ -0,0 +1,145 @@
+	public PermissionGenerator(Builder builder, final String... args) {
+		this.builder = builder;
+
+		if (args.length == 2 && PARAM_ALL.equals(args[1])) {
```
I think we sholud get rid of "all". It is not documented at http://bnd.bndtools.org/macros/permissions.html and was broken in the prior implementation anyways (It was basically an equivalent to "packages"). And this implementation only work if "all" is the sole argument.

marcdejonge replied:
I think this code is equivalent to the previous version, that also only worked if "all" was the only parameter.

But I agree that maybe it should be dropped or well documented.

bjhargrave replied:
In the original code `all` was equivalent to `packages` since each iteration over args would match the first if in the if/else chain. `${permissions; admin; all}` was the same as `${permissions; admin; packages}`.

marcdejonge replied:
Furthermore, I'd often would want to use the capability, packages and services, but not the admin. The admin permission is not often used, but the others are. Maybe create a special parameter for that?

bjhargrave replied:
Lets just remove `all`. It is not helpful since the definition is subject to change as we add new support like now. It is not hard to say `${permissions;packages;services;capabilities}`.

bjhargrave commented on biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java:
```diff
@@ -0,0 +1,145 @@
+				generatePackages();
+			} else if (PARAM_SERVICES.equals(arg)) {
+				generateServices();
+			} else if (KEY.equals(arg))
```
I am not sure what the point of `${permissions; permissions}` since it appears to be basically a noop. I know it was this way before this PR so I am just calling into question the need to continue to support it.

marcdejonge replied:
What I did see while debugging is that for the macro `${permissions;services}` I get 2 parameters, `args[0] = "permissions"` and `args[1] = "services"`.

bjhargrave replied:
Ah. The macro processor must put the macro name in arg[0]. Since the code does a foreach loop on the args array, it needs to ignore element 0 rather then using a for loop with an index starting at 1.

bjhargrave replied:
Since the constructor puts the array in a List, it could remove the first element "permissions" and thus the need for the test in the for loop.

bjhargrave commented on biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java:
```diff
@@ -0,0 +1,145 @@
+	}
+
+	private void generateAdmin() {
+		sb.append("(org.osgi.framework.AdminPermission)");
```
End with `\n`

bjhargrave commented on biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java:
```diff
@@ -0,0 +1,145 @@
+		for (String namespace : builder.getProvideCapability().keySet()) {
+			if (!Processor.isDuplicate(namespace)) {
+				sb.append("(org.osgi.framework.CapabilityPermission \"").append(namespace)
+						.append("\" \"provide\")\r\n");
```
End with `\n`. We shouldn't use `\r\n` and we want a consistent output regardless of whether the bundle is built on *ix or windows.

marcdejonge replied:
I forgot, will fix all of these...

bjhargrave commented on biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java:
```diff
@@ -0,0 +1,145 @@
+		for (String namespace : builder.getRequireCapability().keySet()) {
+			if (!Processor.isDuplicate(namespace)) {
+				sb.append("(org.osgi.framework.CapabilityPermission \"").append(namespace)
+						.append("\" \"require\")\r\n");
```
End with `\n`. We shouldn't use `\r\n` and we want a consistent output regardless of whether the bundle is built on *ix or windows.

bjhargrave commented on biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java:
```diff
@@ -0,0 +1,145 @@
+				if (!imp.isJava()) {
+					sb.append("(org.osgi.framework.PackagePermission \"");
+					sb.append(imp);
+					sb.append("\" \"import\")\r\n");
```
End with `\n`. We shouldn't use `\r\n` and we want a consistent output regardless of whether the bundle is built on *ix or windows.

bjhargrave commented on biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java:
```diff
@@ -0,0 +1,145 @@
+			for (PackageRef exp : builder.getExports().keySet()) {
+				sb.append("(org.osgi.framework.PackagePermission \"");
+				sb.append(exp);
+				sb.append("\" \"export\")\r\n");
```
End with `\n`. We shouldn't use `\r\n` and we want a consistent output regardless of whether the bundle is built on *ix or windows.

bjhargrave commented on biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java:
```diff
@@ -0,0 +1,145 @@
+	private void generateServices() {
+		for (String declaredService : getDeclaredServices()) {
+			sb.append("(org.osgi.framework.ServicePermission \"").append(declaredService)
+					.append("\" \"register\")\r\n");
```
End with `\n`. We shouldn't use `\r\n` and we want a consistent output regardless of whether the bundle is built on *ix or windows.

bjhargrave commented on biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java:
```diff
@@ -0,0 +1,145 @@
+					.append("\" \"register\")\r\n");
+		}
+		for (String referencedService : getReferencedServices()) {
+			sb.append("(org.osgi.framework.ServicePermission \"").append(referencedService).append("\" \"get\")\r\n");
```
End with `\n`. We shouldn't use `\r\n` and we want a consistent output regardless of whether the bundle is built on *ix or windows.

bjhargrave commented on biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java:
```diff
@@ -0,0 +1,145 @@
+				Attrs attrs = entry.getValue();
+				String ifaces = attrs.get("objectClass");
+				if (ifaces != null) {
+					for (String iface : ifaces.split(",")) {
```
Perhaps split with "\\s*,\\s*" to handle whitespace around comma? I wonder is the user specified this in the manifest with whitespace around the commas if bnd would strip that out? Handling it here means we don't care.

bjhargrave commented on biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java:
```diff
@@ -0,0 +1,145 @@
+					if (matcher.matches()) {
+						referencedServices.add(matcher.group(1));
+					} else {
+						builder.warning("Unexpected osgi.service requirement filter: {}", filter);
```
A user could specify a valid filter directive without an objectClass. So we should not warn here. Just ignore.

marcdejonge replied:
I agree that the warning is not needed here. But what should we do if someone adds a wildcard filter?

bjhargrave commented:
Also, please squash your commits together and force push it so there is just the single commit. Thanks.

bjhargrave commented on biz.aQute.bndlib.tests/src/test/component/DSAnnotationTest.java:
```diff
@@ -2774,4 +2809,105 @@ public static void testMixedStandardBnd() throws Exception {
+
+		File tmpFile = File.createTempFile("bndtest", "permissions.perm");
+		FileWriter fileWriter = new FileWriter(tmpFile);
+		fileWriter.write("${permissions;packages}${permissions;services}${permissions;capability}");
```
Should this be `${permissions;packages;services;capabilities}`?

bjhargrave replied:
Should the test not also cover `${permissions;admin}`?

bjhargrave commented on biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java:
```diff
@@ -0,0 +1,145 @@
+
+	public static final String	PARAM_ALL			= "all";
+	public static final String	PARAM_ADMIN			= "admin";
+	public static final String	PARAM_CAPABILITY	= "capability";
```
This should be plural like services and packages. So `capabilities`.

marcdejonge commented:
Ok, I've fixed of lot of these comments and have extended the tests quite a bit. Since there were this many tests now, I've also put that into a separate test class.

Still not sure how we should handle the "all" parameters. Leave it out for now?

bjhargrave commented:
> Still not sure how we should handle the "all" parameters. Leave it out for now?

Yes. Remove the all option entirely.

marcdejonge commented:
Ok, removed in the last amend :smile:

bjhargrave commented on biz.aQute.bndlib/src/aQute/bnd/osgi/PermissionGenerator.java:
```diff
@@ -0,0 +1,147 @@
+					String filter = attrs.get("filter:");
+					if (filter != null && !filter.isEmpty()) {
+					Matcher matcher = FILTER_PATTERN.matcher(filter);
+					if (matcher.matches()) {
```
Since the filter expression can  be arbitrarily complex, we can only try to handle a reasonable subset. Right now you handle the single expression `(objectClass=com.foo.Service)`. We could also expand to handle trailing wildcards in the service name: `(objectClass=com.foo.*)`. A more complex thing to support would be the objectClass expression in an AND list: `(&(X)(objectClass=com.foo.*)(Y))` where X and Y are other expressions. This latter thing would be a fairly complex regex so we can leave out for now.

pkriens replied:
There is a FilterParser in aQute.bnd.osgi.resource that parses and optimizes filters. You could try to use this class since it handles a lot of these cases much better than a simple Regex

marcdejonge commented:
Ok, I've added a more advanced parsing mechanism for the capability filter, although it is still somewhat limited. It can now also parse filters such as:

```
(objectClass=*)
(objectClass=test.*)
(!(objectClass=test.*))
(|(objectClass=class1)(objectClass=class2))
```

And probably some more. But there are some limitations (e.g. what to do with the & operation? Now it generates nothing).

To do this, I needed to add a way to the FilterParser expressions such that it is much easier to do something based on the type of expression. For this I've added the visitor pattern to the expressions.

marcdejonge commented:
btw, why is the appveyor build failing? It is in some part I haven't touched as far as I know.

bjhargrave commented:
> btw, why is the appveyor build failing? It is in some part I haven't touched as far as I know.

We just added a CI build for Windows using Appveyor and there are some windows test failures we need to resolve. The test failures don't appear to be related to your changes.

marcdejonge commented:
Thanks for all your feedback @bjhargrave, you were right that in all 'exotic filters' we simple should not generate anything. Fixed and rebased.
